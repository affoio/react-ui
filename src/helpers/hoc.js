import { createElement } from 'react';
import { omit } from 'ramda';

const componentFromProp = propName => {
  const Component = props => createElement(props[propName], omit([propName], props));
  Component.displayName = `componentFromProp(${propName})`;
  return Component;
};

export default componentFromProp;
