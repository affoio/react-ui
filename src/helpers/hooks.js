import { useEffect, useState } from 'react';

export const useOnClickOutside = (ref, handler) => {
  useEffect(() => {
    const listener = event => {
      // Do nothing if clicking ref's element or descendent elements
      if (!ref.current || ref.current.contains(event.target)) {
        return;
      }

      handler(event);
    };

    document.addEventListener('click', listener);
    document.addEventListener('touchstart', listener);
    document.addEventListener('focus', listener, true);

    return () => {
      document.removeEventListener('click', listener);
      document.removeEventListener('touchstart', listener);
      document.removeEventListener('focus', listener);
    };
  }, [ref, handler]);
};

export const useDebounce = (value, delay) => {
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);

    return () => {
      clearTimeout(handler);
    };
  }, [delay, value]);

  return debouncedValue;
};
