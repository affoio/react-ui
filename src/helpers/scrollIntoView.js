import smoothscroll from 'smoothscroll-polyfill';

smoothscroll.polyfill();

export function scrollIntoView(menuEl, focusedEl) {
  const menuRect = menuEl.getBoundingClientRect();
  const focusedRect = focusedEl.getBoundingClientRect();
  const overScroll = focusedEl.offsetHeight / 3;

  if (focusedRect.bottom + overScroll > menuRect.bottom) {
    menuEl.scrollTo(
      menuEl,
      Math.min(focusedEl.offsetTop + focusedEl.clientHeight - menuEl.offsetHeight + overScroll, menuEl.scrollHeight)
    );
  } else if (focusedRect.top - overScroll < menuRect.top) {
    menuEl.scrollTo(menuEl, Math.max(focusedEl.offsetTop - overScroll, 0));
  }
}
