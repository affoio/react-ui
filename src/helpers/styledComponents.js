import { css } from 'styled-components';

export const addStyleIfPropTruly = (propName, style) => props => {
  if (props[propName]) {
    return css`
      ${style}
    `;
  }
};
