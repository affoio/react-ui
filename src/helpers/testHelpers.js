import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render } from 'enzyme';

import { MemoryRouter } from 'react-router';

export const renderWithThemeProvider = (children, theme = {}) =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);

export const withProviders = (component, initialEntries, theme = {}) => (
  <MemoryRouter initialEntries={initialEntries}>
    <ThemeProvider theme={theme}>{component}</ThemeProvider>
  </MemoryRouter>
);
