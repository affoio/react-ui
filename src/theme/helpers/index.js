import { css } from 'styled-components';

export const isInteger = value => parseFloat(value) === parseInt(value, 10);
export const calculatePercentageGridValue = (value, gridSize) => `${(100 / gridSize) * value}%`;
export const getLimitedMedia = (breakpoints, current, currentIndex) => {
  return currentIndex !== Object.keys(breakpoints).length - 1
    ? `(min-width: ${breakpoints[current]}px) and (max-width: ${breakpoints[
        Object.keys(breakpoints)[currentIndex + 1]
      ] - 1}px)`
    : `(min-width: ${breakpoints[current]}px)`;
};

export const mediaQueryFunctionsList = breakpoints =>
  Object.keys(breakpoints).reduce((media, breakpoint, index) => {
    const newMedia = { ...media };
    newMedia[breakpoint] =
      index === 0
        ? (...args) => css`
            ${css(...args)}
          `
        : (...args) => css`
            @media (min-width: ${breakpoints[breakpoint]}px) {
              ${css(...args)}
            }
          `;
    return newMedia;
  }, {});

export const mediaQueryFunctionsListWithLimits = breakpoints =>
  Object.keys(breakpoints).reduce((media, breakpoint, index) => {
    const newMedia = { ...media };
    newMedia[breakpoint] = (...args) => css`
      @media ${getLimitedMedia(breakpoints, breakpoint, index)} {
        ${css(...args)}
      }
    `;
    return newMedia;
  }, {});
