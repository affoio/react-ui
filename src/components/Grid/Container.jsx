import React from 'react';
import styled, { withTheme } from 'styled-components';
import { bool } from 'prop-types';
import {
  getBaseIndent,
  getBreakpoints,
  getContainerIndents,
  getContainerMaxWidth,
  getGutterWidth,
} from '../../theme/selectors';
import { mediaQueryFunctionsList } from '../../theme/helpers';

const createMediaQueries = (indents, media) => {
  return Object.keys(indents).map(breakpoint => {
    return media[breakpoint]`
        padding-left: ${indents[breakpoint]}px;
        padding-right: ${indents[breakpoint]}px;
      `;
  });
};

const ContainerStyled = styled.div`
  width: 100%;
  box-sizing: border-box;
  max-width: ${({ theme }) => getContainerMaxWidth(theme)};
  margin-left: auto;
  margin-right: auto;

  ${({ fullWidth, mediaQuery, theme }) => {
    const fullWidthPadding = `${getBaseIndent(theme) * getGutterWidth(theme)}px`;
    return fullWidth
      ? `
      padding-left: ${fullWidthPadding};
      padding-right: ${fullWidthPadding};
    `
      : mediaQuery;
  }}
`;

const propTypes = {
  fullWidth: bool,
};

const Container = props => {
  const { children, fullWidth, theme, className } = props;

  const mediaQuery = React.useMemo(() => {
    const baseIndent = getBaseIndent(theme);
    const containerIndents = getContainerIndents(theme);
    const breakPoints = getBreakpoints(theme);
    const containerIndentsWithBase = Object.keys(containerIndents).reduce((indents, current) => {
      const newIndents = { ...indents };
      newIndents[current] = containerIndents[current] * baseIndent;
      return newIndents;
    }, {});

    return createMediaQueries(containerIndentsWithBase, mediaQueryFunctionsList(breakPoints), baseIndent);
  }, [theme]);

  return <ContainerStyled {...{ fullWidth, mediaQuery, theme, className }}>{children}</ContainerStyled>;
};

Container.propTypes = propTypes;

export default withTheme(Container);
