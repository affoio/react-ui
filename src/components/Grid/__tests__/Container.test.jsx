import React from 'react';
import Container from '../Container';
import { renderWithThemeProvider } from '../../../helpers/testHelpers';

describe('<Container>', () => {
  it('renders Container', () => {
    const tree = renderWithThemeProvider(<Container>Test Text</Container>);

    expect(tree).toMatchSnapshot();
  });

  it('renders fullWidth Container', () => {
    const tree = renderWithThemeProvider(<Container fullWidth>Test Text</Container>);

    expect(tree).toMatchSnapshot();
  });

  it('renders Container with theme', () => {
    const tree = renderWithThemeProvider(<Container>Test Text</Container>, {
      baseIndent: 8,
      container: { indent: { xs: 8, sm: 10, lg: 4 } },
    });

    expect(tree).toMatchSnapshot();
  });
});
