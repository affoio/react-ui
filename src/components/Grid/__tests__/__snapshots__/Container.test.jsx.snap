// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Container> renders Container 1`] = `
.c0 {
  width: 100%;
  box-sizing: border-box;
  max-width: 1920px;
  margin-left: auto;
  margin-right: auto;
  padding-left: 16px;
  padding-right: 16px;
}

@media (min-width:768px) {
  .c0 {
    padding-left: 16px;
    padding-right: 16px;
  }
}

@media (min-width:1280px) {
  .c0 {
    padding-left: 60px;
    padding-right: 60px;
  }
}

<div
  class="c0"
>
  Test Text
</div>
`;

exports[`<Container> renders Container with theme 1`] = `
.c0 {
  width: 100%;
  box-sizing: border-box;
  max-width: 1920px;
  margin-left: auto;
  margin-right: auto;
  padding-left: 64px;
  padding-right: 64px;
}

@media (min-width:768px) {
  .c0 {
    padding-left: 80px;
    padding-right: 80px;
  }
}

@media (min-width:1280px) {
  .c0 {
    padding-left: 32px;
    padding-right: 32px;
  }
}

<div
  class="c0"
>
  Test Text
</div>
`;

exports[`<Container> renders fullWidth Container 1`] = `
.c0 {
  width: 100%;
  box-sizing: border-box;
  max-width: 1920px;
  margin-left: auto;
  margin-right: auto;
  padding-left: 8px;
  padding-right: 8px;
}

<div
  class="c0"
>
  Test Text
</div>
`;
